src/enumerable.cr

```diff
@@ -1766,6 +1766,43 @@ module Enumerable(T)
     tally_by { |item| item }
   end
 
+  # Tallies the collection. Accepts a `hash` to count occurrences.
+  # The value corresponding to each element must be an `Int32`.
+  # Returns a `hash` where the keys are the
+  # elements and the values are numbers of elements in the collection
+  # that correspond to the key after transformation by the given block.
+  #
+  # ```
+  # hash = {} of Char => Int32
+  # words = ["Crystal", "Ruby"]
+  # words.each { |word| word.chars.tally_by(hash, &.downcase) }
+  # hash
+  # => {'c' => 1, 'r' => 2, 'y' => 2, 's' => 1, 't' => 1, 'a' => 1, 'l' => 1, 'u' => 1, 'b' => 1}
+  # ```
+  def tally_by(hash : Hash(U, Int32), & : T -> U) : Hash(U, Int32) forall U
```

Comment:
Can the incoming hash be optional? I'm picturing a more common use case of tallying on a single item:

```crystal
word_counts = "Crystal".chars.tally_by(&.downcase)
# => {'c' => 1, 'r' => 1, 'y' => 1, 's' => 1, 't' => 1, 'a' => 1, 'l' => 1}
word_counts = "crystal".chars.tally
# => {'c' => 1, 'r' => 1, 'y' => 1, 's' => 1, 't' => 1, 'a' => 1, 'l' => 1}
```

Reply:
Actually, a hash is optional because Crystal already has `Enumerable#tally` and `Enumerable#tally_by` methods. And they work exactly as you wrote. 

This PR adds support to pass the hash as an argument to the `#tally` method that stores the count of items.

Reply:
Oh!  Sorry about that...I didn't see the method above in GitHub diff.  I guess on that note, the code is arranged `tally_by, tally, tally_by(hash), tally(hash)`...wondering if it should be `tally_by, tally_by(hash), tally, tally(hash)`...not sure what crystal source typically does here.

Reply:
Also wondering if the "non-hash" version of tally_by should just call this tally by, passing a defaulted Hash.  This should DRY up the actual tallying code.

Reply:
Well, yes. Even better. It can be refactored to accept an **optional** hash instead of creating an additional method.

```crystal
def tally_by(hash = Hash(U, Int32).new, & : T -> U) : Hash(U, Int32) forall U
  each_with_object(hash) do |item, hash|
    value = yield item
    count = hash[value]?
    hash[value] = count ? count + 1 : 1
  end
end
```

If the core team is OK with this I will change the code.